decision-records/2025-03-17--otelcol-arbitrary-config-adr.md

```diff
@@ -0,0 +1,230 @@
+
+## Alternative 1: `juju config`
+
+In order to include `processors` or `extensions` we may use `juju config`, like this:
+
+```shell
+juju config otel-col processors_file='@path/to/processors-config.yaml' to_pripelines='metrics'
+```
+
+Let's imagin we want to enable the [`metricsgeneratorprocessor`](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/metricsgenerationprocessor) to the `metrics` pipeline, we need to create a file with like this one:
+
+```yaml
+metricsgenerator:
+    rules:
+        # create pod.cpu.utilized following (pod.cpu.usage / node.cpu.limit)
+        - name: pod.cpu.utilized
+          type: calculate
+          metric1: pod.cpu.usage
+          metric2: node.cpu.limit
+          operation: divide
+
+        # create pod.memory.usage.bytes from pod.memory.usage.megabytes
+        - name: pod.memory.usage.bytes
+          unit: Bytes
+          type: scale
+          metric1: pod.memory.usage.megabytes
+          operation: multiply
+          scale_by: 1048576
+```
+
+Once this config is added, the config file will have a new section `processors` like this:
+
+```yaml
+processors:
+  metricsgenerator:
+      rules:
+          # create pod.cpu.utilized following (pod.cpu.usage / node.cpu.limit)
+          - name: pod.cpu.utilized
+            type: calculate
+            metric1: pod.cpu.usage
+            metric2: node.cpu.limit
+            operation: divide
+
+          # create pod.memory.usage.bytes from pod.memory.usage.megabytes
+          - name: pod.memory.usage.bytes
+            unit: Bytes
+            type: scale
+            metric1: pod.memory.usage.megabytes
+            operation: multiply
+            scale_by: 1048576
+```
+
+and the `service` section will be like this:
+
+```yaml
+service:
+  pipelines:
+    metrics:
+      receivers: [prometheus]
+      processors: [metricsgenerator]
+      exporters: [prometheus]
+
+    logs:
+      receivers: [loki, filelog]
+      exporters: [loki]
+
+    traces:
+      receivers: [otlp]
+      exporters: [tempo]
+```
+
+
+On the other hand of we need to add `extensions` we may execute:
+
+```shell
+juju config otel-col extensions_file='@path/to/extensions-config.yaml'
+```
+
+The `extensions-config.yaml` would contain something like this:
+
+```yaml
+health_check:
+  endpoint: 0.0.0.0:13133
+pprof:
+  endpoint: 0.0.0.0:1777
+```
+
+And the config will have a new `extensions` section like this:
+
+```yaml
+extensions:
+  health_check:
+    endpoint: 0.0.0.0:13133
+  pprof:
+    endpoint: 0.0.0.0:1777
+```
+
+and the `service` section will have a new `extensions` section:
+
+```yaml
+service:
+  extensions: [health_check, pprof]
+  pipelines:
+    metrics:
+      receivers: [prometheus]
+      processors: [metricsgenerator]
+      exporters: [prometheus]
+
+    logs:
+      receivers: [loki, filelog]
+      exporters: [loki]
+
+    traces:
+      receivers: [otlp]
+      exporters: [tempo]
+```
+
+## Alternative 2: juju actions
+
```

Comment:
cons: the charm has to persist and maintain its own state; in the other cases juju will do that